### ISSUE TITLE
Ladybug, Honeybee, Dragonfly for Sverchok + Blender

### ISSUE THREAD

**Issue by Moult** [nortikin/sverchok#3556]
Just creating an issue to keep track of the progress of creating Ladybug Tools (including Honeybee and Dragonfly) for Sverchok. Note that the Ladybug nodes currently available in Sverchok is not the full package - so the current nodes in Sverchok related to Ladybug needs to be removed, as the actual Ladybug has specific requirements on inputs and outputs. Ladybug tools is a collection of hundreds of nodes that allow solar, energy, and environmental analysis to be done in Blender. It is quite famous in the Grasshopper / Dynamo platform.

I think that at the end these nodes will be shipped separately to Sverchok, because otherwise Sverchok would get too huge out of the box. Like another Sverchok-extras repo. I will be happy to maintain this extension to Sverchok.

Steps:

 1. ~~Discuss with Ladybug tools creators on the process of porting to Blender to ensure upstream compatibility~~ Done
 2. ~~Create prototype which demonstrates non graphical nodes can be ported~~ Done
 3. ~~Create prototype which demonstrates graphical nodes can be ported~~ Done
 4. ~~Obtain JSON node descriptions of the Ladybug suite~~ Done
 5. ~~Generate Sverchok nodes from JSON~~ Done
 6. ~~Combine generated Sverchok nodes with patched Ladybug codebase~~ Done
 7. ~~Basic test with node combinations and build geometry interfaces~~ Done
 8. ~~Package for users (Need to wait for upstream to stabilise first)~~ Done
 9. Test. Test. Test.
 10. Honeybee! Dragonfly!

Screenshot of non-graphical prototype done so far, just to see how it might work best:

![image](https://user-images.githubusercontent.com/88302/93470689-9cf36080-f935-11ea-864f-e85a69f34183.png)

**Comment by nortikin** [comment]
we already have couple of ladybug nodes, and dependency covered with buttons by pip install of external packages.

**Comment by zeffii** [comment]
the number, if you are outputting a single number `[[25]]`
and a single string of text `[["Mascot Sydney Airport ....... elev 5"]]`

**Comment by zeffii** [comment]
looks like your location object has it's own __repr__ function?  :)

**Comment by Moult** [comment]
@nortikin yeah sorry about the confusion, would you mind removing those couple of existing Ladybug nodes? Ladybug has a set of requirements of how nodes are to be created, so those two nodes are actually not part of the original Ladybug. Should I create a pull request to remove them?

@zeffii you are correct about the nesting, it was just a test, but thanks for the reminder! :)

**Comment by nortikin** [comment]
> @nortikin yeah sorry about the confusion, would you mind removing those couple of existing Ladybug nodes? Ladybug has a set of requirements of how nodes are to be created, so those two nodes are actually not part of the original Ladybug. Should I create a pull request to remove them?
> 
> @zeffii you are correct about the nesting, it was just a test, but thanks for the reminder! :)

pending invite

**Comment by vicdoval** [comment]
Awesome! When I was coding the ladybug nodes I was thinking in something like this, but could not find the time/knowledge...

If you are sure with the external repository solution I would like to fix #3359 (I have to take it back to mi mind it was a bit buried)

On my view the actual Ladybug nodes and dependencies can be removed when you have the repository ready to be used, let me do it when the moment arrives.

So please keep me updated about your progress and thanks for doing this!

**Comment by zeffii** [comment]
@Moult  there is also 

- a `debug_print` node, which takes the output from any connected socket and dumps it into the terminal, this can be useful for ..well.... debugging :) - it doesn't care what you pipe into it.
- `stethoscope` node, you've already encountered, but it also isn't very particular about the data you feed into it.

**Comment by nortikin** [comment]
> In Grasshopper (and I assume Dynamo), many nodes will generate un-baked geometry when run. In Sverchok, I think the paradigm is different, and basically the data is only operated on with no concept of geometry (i.e you don't pass around mathutils.Vector, and you don't pass around meshes). So the only way to view the result is to connect it to a viewer node. Is this correct, or is there another approach (i.e. a way for any node to call the capabilities of the viewer nodes themselves)? And the viewer node you connect to depends on whether it is text (#3559), mesh, curve, etc, correct?
> 
viewer draw is openGL
viewer bmesh is autobake
viewers polyline/others baking also different objects
viewer 2d opengl in node editor
text viewer/sthetoscope/debug print/others text display, even note node.

**Comment by portnov** [comment]
Contributing.md should be updated with respect to API changes... Methods it refers to mostly still exist, but there are newer and better in many cases.

**Comment by vicdoval** [comment]
> Is there a place I can catch you online for live chat? 

I have send you a email with my contact details :)

**Comment by Moult** [comment]
Thanks @vicdoval ! I will try not to bother you :)

Here is a bit of progress - a sunpath being generated based off location data in an EPW file. By the way, has anybody written any code for interpolation between a list of 3D vertices? It is needed in the Ladybug code :)

Although these nodes are manually ported for now, I have manually ported them in a way so that later on I can bulk import all the nodes and it should Just Work (TM), so it is showing promising signs.

![image](https://user-images.githubusercontent.com/88302/93666339-c42a6900-fac0-11ea-925a-ec17cb820f64.png)

**Comment by portnov** [comment]
There are several implementations of interpolation in Sverchok. For simplest ones, look at sverchok.utils.geom.LinearSpline and CubicSpline, and their usages.

**Comment by Moult** [comment]
Thanks @portnov ! I will check it.

Hacky work-in-progress code is now available in these two repos:

https://github.com/Moult/ladybug_tools
https://github.com/Moult/ladybug_blender

**Comment by Moult** [comment]
More progress - 2D geometry, and translation of Sverchok vectors into Ladybug geometry:

![image](https://user-images.githubusercontent.com/88302/93693761-6856f300-fb47-11ea-97ae-87fce55747b4.png)

**Comment by Moult** [comment]
Coloured mesh output now possible. Improved viewer node. Full node set for Ladybug is now generated, with support for default values, list vs item access, and data types.

![image](https://user-images.githubusercontent.com/88302/93762451-eeac2b80-fc52-11ea-928b-54fc1556cd93.png)

**Comment by Moult** [comment]
More nodes! More data! More viz!

![image](https://user-images.githubusercontent.com/88302/93835841-f0fa9e00-fcc3-11ea-8dd7-8ab57fc04bf5.png)

![image](https://user-images.githubusercontent.com/88302/93835847-f657e880-fcc3-11ea-9486-d3e90ba47711.png)

**Comment by nortikin** [comment]
consistant naming that what i love.
LB out instead of ladybug to sverchok
LB prefix is great.

Hope i could some day make refactoring with consistancy for polygons/edges/vertices/numbers/colors and other sockets and nodes. it will be great day.

**Comment by nortikin** [comment]
Ladybug and Sverchok marriage

**Comment by Moult** [comment]
Ray intersection is now implemented. Check out the first results of simulation!

If you ever wondered the impact of sun hours on the 21st of June in Sydney of a building shaped like a monkey, now you know.

![image](https://user-images.githubusercontent.com/88302/94118359-c9a4fc00-fe90-11ea-8fea-735dc9e1326d.png)

**Comment by JitQL** [comment]
This is great work! Sverchok and Ladybugs is really a game changer for people like me.

**Comment by Spirou4D** [comment]
@nortikin Congratulations! Wonderfull job! ;-))

**Comment by onerawartist** [comment]
Amazing work ! Blender is going BIM 😍 this this history that im witnessing ^^

**Comment by Moult** [comment]
Repo with basic makefile (`make dist`) for packaging now available: https://github.com/ladybug-tools/ladybug-blender

**Comment by ppaawweeuu** [comment]
Can you write Installation chapter, please.. It seems that downloading zipped code from 'ladybug-blender' folder & standard 'Install Addon' procedure does't work (or is not enough). thx

**Comment by nortikin** [comment]
> @nortikin Congratulations! Wonderfull job! ;-))

all thanks to @Moult

**Comment by Moult** [comment]
@ppaawweeuu correct, it is not enough. There are two steps:

 1. https://github.com/nortikin/sverchok/pull/3565 needs to be merged
 2. `make dist` needs to be run to generate the add-on zip file

Once it is merged, I will provide a downloadable zip which people can start testing with. Sorry for the wait!

**Comment by Moult** [comment]
@ppaawweeuu first release **v0.0.200927** is out! Check it out here: https://community.osarch.org/discussion/comment/2910/#Comment_2910

**Comment by nortikin** [comment]
![изображение](https://user-images.githubusercontent.com/5783432/109402978-63566500-796b-11eb-8904-26dbe579476e.png)
[LB.blend.zip](https://github.com/nortikin/sverchok/files/6055868/LB.blend.zip)

**Comment by nortikin** [comment]
is there github repo for LB and IFC for sverchok?

**Comment by Moult** [comment]
https://github.com/ladybug-tools/ladybug-blender
https://github.com/IfcOpenShell/IfcOpenShell/tree/v0.6.0/src/ifcsverchok

**Comment by Durman** [comment]
Probably we could add list of add-ons for Sverchok in main github page.

**Comment by vicdoval** [comment]
I added it to the wiki some days ago https://github.com/nortikin/sverchok/wiki/Sverchok-Extensions

**Comment by vicdoval** [comment]
Also i was thinking about the  possibility of installing them from Sverchok preferences

**Comment by nortikin** [comment]
> Also i was thinking about the possibility of installing them from Sverchok preferences

than we need 'sh' and 'bin' installers to make topologicPY and other dependencies.
https://github.com/wassimj/topologicsverchok/issues/1 here
